provider/pkg/provider/remote/connection.go

```diff
@@ -114,48 +120,95 @@ func (con *Connection) SShConfig() (*ssh.ClientConfig, error) {
 	return config, nil
 }
 
+func dialWithRetry[T any](ctx p.Context, msg string, maxAttemptes int, f func() (T, error)) (T, error) {
+	var userError error
+	_, data, err := retry.Until(ctx, retry.Acceptor{
+		Accept: func(try int, _ time.Duration) (bool, any, error) {
+			var result T
+			result, userError = f()
+			if userError == nil {
+				return true, result, nil
+			}
+			dials := try + 1
+			if reachedDialingErrorLimit(dials, maxAttemptes) {
+				return true, nil, userError
```

Comment:
You might want to pack some information like "After N attempts" into the userError.